lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/driver/ContainerManager.java

```diff
@@ -134,43 +161,227 @@ public void onNext(final Time value) {
+
+  /**
+   * Returns the node name of the container to be allocated if it's available, selected from the list of preferred
+   * node names. If the list is empty, then an empty optional is returned
+   *
+   * @param rackNames
+   *          the list of preferred racks
+   * @return the node name where to allocate the container
+   */
+  private Optional<String> getPreferredNode(final List<String> nodeNames) {
+    if (CollectionUtils.isNotEmpty(nodeNames)) {
+      for (final String nodeName : nodeNames) {
+        final String possibleRack = racksPerNode.get(nodeName);
+        if (possibleRack != null
+            && freeNodesPerRack.get(possibleRack).containsKey(nodeName)) {
+          return Optional.of(nodeName);
+        }
+      }
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Returns the rack where to allocate the container, selected from the list of
+   * preferred rack names. If the list is empty, and there's space in the default
+   * rack, then the default rack is returned. The relax locality semantic is
+   * enabled if the list of rack names contains '/*', otherwise relax locality
+   * is considered disabled
+   *
+   * @param rackNames
+   *          the list of preferred racks
+   * @return the rack name where to allocate the container
+   */
+  private Optional<String> getPreferredRack(final List<String> rackNames) {
+    final Collection<String> normalized = getRackNamesOrDefault(rackNames);
+    for (final String rackName : normalized) {
+      // if it does not end with the any modifier,
+      // then we should do an exact match
+      if (!rackName.endsWith(ANY)) {
+        if (freeNodesPerRack.containsKey(rackName)
+            && freeNodesPerRack.get(rackName).size() > 0) {
+          return Optional.of(rackName);
+        }
+      } else {
+        // if ends with the any modifier, we do a prefix match
+        final Iterator<String> it = availableRacks.iterator();
+        while (it.hasNext()) {
+          final String possibleRackName = it.next();
+          // remove the any modifier
+          final String newRackName = rackName.substring(0,
+              rackName.length() - 1);
+          if (possibleRackName.startsWith(newRackName)) {
+            if (freeNodesPerRack.get(possibleRackName).size() > 0) {
+              return Optional.of(possibleRackName);
+            }
+          }
+        }
+      }
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Allocates a container based on a request event. First it tries to match a
+   * given node, if it cannot, it tries to get a spot in a rack
+   *
+   * @param requestEvent
+   *          the request event
+   * @return an optional with the container if allocated
+   */
+  Optional<Container> allocateContainer(final ResourceRequestEvent requestEvent) {
+    Container container = null;
+    final Optional<String> nodeName = getPreferredNode(requestEvent
+        .getNodeNameList());
+    if (nodeName.isPresent()) {
```

Comment:
We first try for the node name... if it's taken, then we fallback to the rack names...